toolkit/tools/imagecustomizerlib/imagecustomizer.go

```diff
@@ -0,0 +1,128 @@
+) error {
+	var err error
+
+	var config imagecustomizerapi.SystemConfig
+	err = imagecustomizerapi.UnmarshalYamlFile(configFile, &config)
+	if err != nil {
+		return err
+	}
+
+	baseConfigPath, _ := filepath.Split(configFile)
+
+	err = CustomizeImage(buildDir, baseConfigPath, &config, imageFile, outputImageFile, outputImageFormat)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func CustomizeImage(buildDir string, baseConfigPath string, config *imagecustomizerapi.SystemConfig, imageFile string,
+	outputImageFile string, outputImageFormat string,
+) error {
+	var err error
+
+	imageCustomizerDir := filepath.Join(buildDir, "imagecustomizer")
+
+	err = os.MkdirAll(imageCustomizerDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	// Convert image file to raw format, so that a kernel loop device can be used to make changes to the image.
+	buildImageFile := filepath.Join(imageCustomizerDir, "image.raw")
+
+	_, _, err = shell.Execute("qemu-img", "convert", "-O", "raw", imageFile, buildImageFile)
+	if err != nil {
+		return fmt.Errorf("failed to convert image file to raw format: %w", err)
+	}
+
+	// Customize the raw image file.
+	err = customizeImageHelper(buildDir, baseConfigPath, config, buildImageFile)
+	if err != nil {
+		return err
+	}
+
+	// Create final output image file.
+	_, _, err = shell.Execute("qemu-img", "convert", "-O", toQemuImageFormat(outputImageFormat), buildImageFile, outputImageFile)
+	if err != nil {
+		return fmt.Errorf("failed to convert image file to format: %s: %w", outputImageFormat, err)
+	}
+
+	return nil
+}
+
+func toQemuImageFormat(imageFormat string) string {
+	switch imageFormat {
+	case "vhd":
+		return "vpc"
+
+	default:
+		return imageFormat
+	}
+}
+
+func customizeImageHelper(buildDir string, baseConfigPath string, config *imagecustomizerapi.SystemConfig,
+	buildImageFile string,
+) error {
+	// Mount the raw disk image file.
+	diskDevPath, err := diskutils.SetupLoopbackDevice(buildImageFile)
+	if err != nil {
+		return fmt.Errorf("failed to mount raw disk (%s) as a loopback device: %w", buildImageFile, err)
+	}
+	defer diskutils.DetachLoopbackDevice(diskDevPath)
+
+	// Look for all the partitions on the image.
+	newMountDirectories, mountPoints, err := findPartitions(diskDevPath)
+	if err != nil {
+		return err
+	}
+
+	// Create chroot environment.
+	imageChrootDir := filepath.Join(buildDir, "imageroot")
+
+	imageChroot := safechroot.NewChroot(imageChrootDir, false)
+	err = imageChroot.Initialize("", newMountDirectories, mountPoints)
+	if err != nil {
+		return err
+	}
+	defer imageChroot.Close(false)
+
+	// Do the actual customizations.
+	err = doCustomizations(baseConfigPath, config, imageChroot)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func findPartitions(diskDevice string) ([]string, []*safechroot.MountPoint, error) {
+	newMountDirectories := []string{}
+
+	// TODO: Dynamically find partitions instead of hardcoding the mappings.
```

Comment:
Any idea on how you want to approach this?

Reply:
Yes, I have written a [change](https://github.com/cwize1/CBL-Mariner/blob/1017770c0ff566dbde31eba43be012ba57a6db75/toolkit/tools/imagecustomizerlib/imagecustomizer.go#L125C9-L125C9) for this.

I temporarily mount the EFI system partition, read the `grub.cfg`, and pull out the UUID of the OS partition. (The logic currently doesn't support layouts that have an EFS, boot, and OS partition. But I can add support for that later.)

Reply:
Has there been talk of using [discoverable partition types](https://uapi-group.org/specifications/specs/discoverable_partitions_specification/)? That would avoid having to parse grub.cfg or /etc/fstab

Reply:
Does anyone actually use the "Discoverable Partitions Specification". Or is it just another instance of [xkcd 927](https://xkcd.com/927)?

Reply:
A bunch of the systemd utilities integrate with them. I think many of the proposals around unified kernel images (UKIs) also rely on them so that the root partition can be inferred from the disk, rather than passed in as a kernel cmdline argument. Not sure whether there's been much/any adoption by other distros at this point

Reply:
If Mariner ever supports this scheme, we can probably add support for it. Though we would probably still fall back to reading the grub.cfg and fstab files in case the image that doesn't support it.

Reply:
@hbeberman FYI